provider/mem/mem.go

```diff
@@ -200,6 +222,7 @@ func (a *Alerts) Put(alerts ...*types.Alert) error {
 	for _, alert := range alerts {
 		fp := alert.Fingerprint()
 
+		a.mtx.Lock()
```

Comment:
I think to make the `Put` operation atomic we should acquire the lock at the start of the method and then release it with `defer`?

Reply:
moving to the start seems the same to me, I believe the original intent is to make the lock scope as small as possible

Reply:
however, moving to the start is better as this simplifies the logic.

Reply:
> moving to the start seems the same to me

It's not quite the same. The difference is if we move it to the start then the entire Put operation is atomic. This stops two goroutines from interleaving their Put operations.

For example, goroutines 1 puts two alerts:

```
labels={foo=bar},status=firing
labels={bar=baz},status=resolved
```

And goroutine 2 puts the same alerts, but with swapped statuses:

```
labels={foo=bar},status=resolved
labels={bar=baz},status=firing
```

The order of insertion between the two goroutines is random and unpredictable. But if we acquire the lock at the start we make the entire operation serializable.

Reply:
I believe this makes sense, but the difference in a batch isn't that significant either.

Reply:
this lock may contains CPU-sensitive operations such as fingerprint calculations